python/hsfs/core/vector_db_client.py

```diff
@@ -141,8 +142,28 @@ def find_neighbors(
             embedding_feature.embedding_index.col_prefix
             and len(results["hits"]["hits"]) != k
         ):
-            query["query"]["bool"]["must"][0]["knn"][col_name]["k"] = 3 * k
+            # Get the max number of results allowed to request
+            if not self._index_result_limit_k.get(index_name):
+                query["query"]["bool"]["must"][0]["knn"][col_name]["k"] = 2**31 - 1
+                try:
+                    self._opensearch_client.search(body=query, index=index_name)
+                except VectorDatabaseException as e:
+                    if (
+                        e.reason == VectorDatabaseException.REQUESTED_K_TOO_LARGE
+                        and e.info.get(
+                            VectorDatabaseException.REQUESTED_K_TOO_LARGE_INFO_K
+                        )
+                    ):
+                        self._index_result_limit_k[index_name] = e.info.get(
+                            VectorDatabaseException.REQUESTED_K_TOO_LARGE_INFO_K
+                        )
+                    else:
+                        raise e
+            query["query"]["bool"]["must"][0]["knn"][col_name]["k"] = min(
+                self._index_result_limit_k.get(index_name, k), 3 * k
+            )
             results = self._opensearch_client.search(body=query, index=index_name)
+
```

Comment:
shouldnt this be in else statement or somehow different? it looks like as long as there are no exceptions we would always call `self._opensearch_client.search(body=query, index=index_name)` 2X but care only for the second result.

Reply:
The first call always returns exception. I have added comments.